Mutagen.Bethesda.Analyzers/AnalyzerRunnerBuilder.cs

```diff
@@ -1,90 +1,210 @@
     }
 
-    public AnalyzerRunnerBuilder WithFileSystem(IFileSystem fileSystem) {
-        _fileSystem = fileSystem;
-        return this;
+    public AnalyzerRunnerBuilder WithFileSystem(IFileSystem fileSystem)
+    {
+        return this with
+        {
+            _fileSystem = fileSystem
+        };
     }
 
-    public AnalyzerRunnerBuilder WithThreads(int threads) {
-        _numWorkThreadsController = new NumWorkThreadsConstant(threads);
-        return this;
+    public AnalyzerRunnerBuilder WithThreads(int threads)
+    {
+        return this with
+        {
+            _numWorkThreadsController = new NumWorkThreadsConstant(threads)
+        };
+    }
+
+    public AnalyzerRunnerBuilder WithThreads(IObservable<int?> threads)
+    {
+        return this with
+        {
+            _numWorkThreadsController = new NumWorkThreadsByObservable(threads)
+        };
     }
 
-    public AnalyzerRunnerBuilder WithThreads(IObservable<int> threads)
+    private class NumWorkThreadsByObservable(IObservable<int?> numThreads) : INumWorkThreadsController
     {
-        // TODO: replace with implementation accepting an observable
-        _numWorkThreadsController = new NumWorkThreadsConstant(-1);
-        return this;
+        public IObservable<int?> NumDesiredThreads { get; } = numThreads;
     }
 
     public AnalyzerRunnerBuilder WithMinimumSeverity(Severity minimumSeverity)
     {
-        _minimumSeverity = minimumSeverity;
-        return this;
+        return this with
+        {
+            _minimumSeverity = minimumSeverity
+        };
     }
 
     public AnalyzerRunnerBuilder WithTopicConfig(TopicConfig? topicConfig)
     {
-        _topicConfig = topicConfig;
-        return this;
+        return this with
+        {
+            _topicConfig = topicConfig
+        };
+    }
+
+    public AnalyzerRunnerBuilder WithDataDirectory(DirectoryPath dataDirectory)
+    {
+        return this with
+        {
+            _dataDirectory = _dataDirectory
+        };
     }
 
     public IAnalyzerRunner Build()
     {
-        return new AnalyzerRunner(
-            _fileSystem ?? new FileSystem(),
-            _gameRelease,
-            _linkCache,
-            _loadOrder,
-            _topicConfig ?? new TopicConfig(),
-            _minimumSeverity,
-            _numWorkThreadsController ?? new NumWorkThreadsConstant(null));
+        var builder = new ContainerBuilder();
+
+        builder.RegisterModule<MainModule>();
+        builder.RegisterModule<SkyrimAnalyzerModule>();
+
+        builder
+            .RegisterInstance(_fileSystem ?? new FileSystem())
+            .As<IFileSystem>();
+
+        builder
+            .RegisterInstance(new GameReleaseInjection(_gameRelease))
+            .AsImplementedInterfaces();
+
+        builder
+            .RegisterInstance(_loadOrder)
+            .AsImplementedInterfaces();
+
+        builder
+            .RegisterInstance(_linkCache)
```

Comment:
In case we want api side cache validation we'll need to be cautious about this stuff because we'd need a different container with a different instance when this has changed.

Reply:
Can you elaborate on what you mean a bit?   Right now we're offering a builder where the user can give the linkcache to be used.   If the linkcache changes (because the input mods change), then a new builder given a new linkcache will be needed.  I think that's just how it is with the API we're offering with this builder?

Reply:
I think for now we're requiring people to redo their runner every time.    I think if we offer some crazy dynamic system in the future, we can just offer a new builder options that takes in IObservable of something and change the internals appropriately?